sources/core/Xenko.Core/ThreadThrottler.cs

```diff
@@ -0,0 +1,217 @@
+            }
+
+            if (sleepMinDuration >= 1)
+            {
+                long sleepStart = Stopwatch.GetTimestamp();
+                // Not sure about the use of Utilities here, what are the differences between it and Threads',
+                // that should be specified into Utilities.Sleep's summary.
+                Utilities.Sleep(sleepMinDuration);
+                var sleepElapsed = Utilities.ConvertRawToTimestamp(Stopwatch.GetTimestamp() - sleepStart);
+                // Decrease next sleep duration since we slept too much this time, always happens
+                fractionalSleep -= sleepElapsed - TimeSpan.FromMilliseconds(sleepMinDuration);
+
+                finalElapsedTime = callWatch.Elapsed;
+                callWatch.Restart();
+                return true;
+            }
+
+            finalElapsedTime = callWatch.Elapsed;
+            callWatch.Restart();
+            return false;
+        }
+
+        bool ThrottleThreadPrecise(out TimeSpan finalElapsedTime)
+        {
+            if (callWatch == null)
+                callWatch = Stopwatch.StartNew();
+
+            // Throttle only when we are too fast
+            if (callWatch.Elapsed > MinimumElapsedTime)
+            {
+                finalElapsedTime = callWatch.Elapsed;
+                callWatch.Restart();
+                return false;
+            }
+
+            bool looped = false;
+            var oneMs = TimeSpan.FromMilliseconds(1);
+            while (MinimumElapsedTime - callWatch.Elapsed > oneMs + Utilities.ConvertRawToTimestamp(spinwaitWindow))
+            {
+                if (Type == ThrottlerType.PreciseAuto)
+                {
+                    long sleepStart = Stopwatch.GetTimestamp();
+                    Utilities.Sleep(1);
+                    // Include excessive time sleep took on top of the time we specified
+                    spinwaitWindow += (Stopwatch.GetTimestamp() - sleepStart);
+                    // Average to account for general system responsiveness
+                    spinwaitWindow = spinwaitWindow == 0 ? 0 : spinwaitWindow / 2;
+                }
+                else
+                {
+                    Utilities.Sleep(1);
+                }
+                looped = true;
+            }
+
+            // We skipped the loop, probably because the system is/was extremely busy, 
+            // reduce delay slowly to force loop re-entry and re-evaluation of the current averageSleepDelay
+            if (looped == false && Type == ThrottlerType.PreciseAuto)
+                spinwaitWindow = spinwaitWindow == 0 ? 0 : spinwaitWindow - spinwaitWindow / 10;
+
+            // prefer one tick too soon than two ticks too late
+            var epsilon = TimeSpan.FromTicks(1);
+            // 'spinwait' for the rest of the duration, that duration should take close to threadSleepDurationMs
+            while (MinimumElapsedTime - callWatch.Elapsed > epsilon) { }
```

Comment:
Wouldn't it be better to use `System.Threading.SpinWait` instead of manual aggressive empty loop?

Reply:
The class defaults to the other mode, this one's point is to be as precise as possible even to the point of wasting cycles.
Spinwait will yield in certain cases so we won't be able to guarantee users that our timing is precise.

Reply:
But it is very important to yield after some time, otherwise the system might become unstable (CPU starvation), hence why this struct exist in the .Net framework. In the above code, there is no check to ensure that `MinimumElapsedTime - callWatch.Elapsed` is not a very big number, which I find a bit dangerous.

Reply:
We do yield, Utilities.Sleep yields for 1+ms and that function is called multiple times within this function until the time left to sleep is just slightly larger than one ms. 
if `MinimumElapsedTime - callWatch.Elapsed` is large, that's expected behavior, Utilities.Sleep(Thread.Sleep) will just run a bunch more ; what might be worrying is the `spinwaitWindow` scaling up and ignoring the sleep loop but it will always auto-correct itself back the next couple of calls.

We didn't yield before this PR and we still won't yield if we're running slowly, if yielding is necessary it shouldn't lie in this function's implementation since this one's about capping the maximum execution rate not managing load balancing.